### ISSUE TITLE
Trying to understand OnTurnAsync for saving Accesors

### ISSUE THREAD

**Issue by DanteNahuel** [MicrosoftDocs/bot-docs#734]
So I've been posting a bunch of questions but after getting the answer to them I noticed my real problem is that I'm not understanding how does the turn handler works and how do you use the accessors globally over all your project code.

Here's the deal:
Using Reuse Dialogs documentation, I'm trying to build a bot with multiple nested Dialogs that are called from a main Dialog. These Dialogs collect multiple information from user input and (should) save this data using the accessors either to get or set it.

I'm able to accomplish this when working with only one Dialog, but when I start to nest more Dialogs things get complicated and I loose the ability (and the understanding) to get or set accessors.

The main algorithm works like this:
Main Dialog ask the user what does she/he wants:
1) check ticket data
2) check some questions
3) something else

If the user chooses 1), a ticket dialog (which is in a separate .cs) is called.

In this new dialog, the user is prompted for the ticket number and uses some logic to separate the ticket numbers from the ticket letters, and check if the ticket number is valid (custom and own created functions).
Here things start to get complicated.
If the ticket is valid, then a new nested Dialog (in another separated .cs) is called, and this new Dialog will use the ticket number and letters to search it over a database and reply the information to the user.

For this to work, I wanted to use accessors to save the ticket numbers and letters and send them over the Context to the new Dialog to be used against the Database.
However I don't understand how to save these accessors so the information is available globally.

I believe I can use the turn handler OnAsync, since this is checked every turn, to save the info,
but:
In the Reuse Dialogs example, the code saves the Guest information over the Main Dialog, not in the OnTurnAsync using this:

```
    switch (stepContext.Result)
    {
        case TableInfo table:
            // Store the results of the reserve-table dialog.
            userInfo.Table = table;
            await _accessors.UserInfoAccessor.SetAsync(stepContext.Context, userInfo, cancellationToken);
            break;
        case WakeUpInfo alarm:
            // Store the results of the set-wake-up-call dialog.
            userInfo.WakeUp = alarm;
            await _accessors.UserInfoAccessor.SetAsync(stepContext.Context, userInfo, cancellationToken);
            break;
        default:
            // We shouldn't get here, since these are no other branches that get this far.
            break;
    }
```

But this happens only after the previous step is finished so it's no good for me.

The OnTurnAsync has this example for saving Accessors:

```
// Process the result of any complete dialog.
        if (dialogTurnResult.Status is DialogTurnStatus.Complete)
        {
            switch (dialogTurnResult.Result)
            {
                case GuestInfo guestInfo:
                    // Store the results of the check-in dialog.
                    userInfo.Guest = guestInfo;
                    await _accessors.UserInfoAccessor.SetAsync(turnContext, userInfo, cancellationToken);
                    break;
                default:
                    // We shouldn't get here, since the main dialog is designed to loop.
                    break;
            }
        }
```

But my code hasn't met this condition (DialogTurnStatus.Complete) so far, nor do I understand how from a running dialog I can send a .Result back to the OnTurnAsync.

So, basically:

What can I do in a waterfall step over a running Dialog to signal the OnTurnAsync to save the Accessors?

When a waterfall step is over and the next one starts, does the OnTurnAsync receives any kind of .result so I can apply a switch
```
case (TicketNumber ticket)
userInfo.TicketNumber = ticket;
                        await _accessors.UserInfoAccessor.SetAsync(stepContext.Context, userInfo, cancellationToken);
```

The only similar thing I can think of is this:
```
return await step.EndDialogAsync(
            (TicketNumber)step.Values[ticketnumber],
            cancellationToken);

```

But this Ends the current Dialog and that's not what I want.
I want the value to be saved when going from one waterfall step to the next one.

Anyway sorry for the long text and thanks in advance.

**Comment by JonathanFingold** [comment]
Here is some conceptual material in the docs that might help you get oriented:
* About [managing state](https://docs.microsoft.com/azure/bot-service/bot-builder-concept-state)
* The [waterfall step context properties](https://docs.microsoft.com/azure/bot-service/bot-builder-concept-dialog#waterfall-step-context-properties) and [using dialogs](https://docs.microsoft.com/azure/bot-service/bot-builder-concept-dialog#using-dialogs) sections of the **dialog libraries** topic

> For this to work, I wanted to use accessors to save the ticket numbers and letters and send them over the Context to the new Dialog to be used against the Database.
> However I don't understand how to save these accessors so the information is available globally.

- Use the step context's **Values** property to collect and aggregate input in a waterfall dialog.
- Use the **options** parameter to pass information to a dialog when you begin it.
- Use the **result** parameter to pass information to the calling context when you end a dialog.
  - If the calling context was a waterfall step, the result is available in the following step via the step context's **Result** property.
  - If the calling context was the turn handler, the result is available in the dialog turn result's **Result** property (when the dialog turn status is **Complete**).

All of this state is managed via the DialogState accessor you provided when creating the outermost DialogSet. The upside to this is that if you cancel a dialog, the associated state goes away and in a sense is never committed.

If you need to access other bot state objects, you can do so, but you'll need to be careful about how you deal with interruptions to the dialog flow. When defining a component dialog, you can require other state property accessors in its constructor that you would then use to manage this additional state from within the dialog. Other options abound, each with their own gotchas.

Let me know if you need more info about any particular aspect of this. Also, , you can post questions at [Stack Overflow](https://stackoverflow.com/questions/tagged/botframework) using the `botframework` tag. Please note that Stack Overflow has guidelines such as requiring a descriptive title, a complete and concise problem statement, and sufficient details to reproduce your issue. Feature requests or overly broad questions are off-topic; new users should visit the [Stack Overflow Help Center](https://stackoverflow.com/help/how-to-ask) for more details.

**Comment by DanteNahuel** [comment]
Thank you for all your helping.
I know of Stack and I find it very useful for things like Arduino programming or other hobbies. However the c# community there is much more aggressive and replies with the game-equal to "git gud" when I ask these kind of questions or problems, even if I try my best to explain my issue with as many details as I can.


Those definitions helped me understand how to pass through info between dialogs or steps. Thanks again.

However I still don't understand how can I save info in between turns using the OnTurnAsync.

Here's my understanding so far:

- I prompt the user for numbers.
- I then save these numbers as step.Values like this:

```
        // Save the ticket number as a string
        string resultString = step.Result as string;

        var userInfo = step.Options;
        //Retrieve the resultString letters and numbers
        string ticketLetters = string.Join("", resultString.ToLower().ToLower().ToCharArray().Where(Char.IsLetter));
        string ticketNumbers = string.Join("", resultString.ToLower().ToCharArray().Where(Char.IsDigit));



        //Save ticket number and letters as properties of TicketNumber State. We need to check if this can be used for something, 
        //since we are not returning it to the main dialog and thus it is not saved in the accessor
        ((TicketNumber)step.Values[ticketKey]).ticketNumber = ticketNumbers;
        ((TicketNumber)step.Values[ticketKey]).ticketLetter = ticketLetters;

```



- And then I call a new nested Dialog (in a different .cs file) like this:

`return await step.BeginDialogAsync(ticketProcessingDialogId, step.Values, cancellationToken);`

- 
And I know I receive it in the other end using this:

`var key = stepContext.Options;`




But all this is inside these Dialogs. I'm not really saving the User State Accessor properties, because as you said it, once the Dialog ends, the state goes away and is not commited.

When I check the Reuse Dialog Documentation examples, I can see that once the **HandleChoiceAsync** step and thus, the nested Dialog ends, it returns the User input as a **step.Result** to the next waterfall step, the **LoopBackAsync**, and in here, depending on the Returned object class, it saves the info in the accessor for good like this:

```
switch (stepContext.Result)
            {
                case TableInfo table:
                    // Store the results of the reserve-table dialog.
                    userInfo.Table = table;
                    await _accessors.UserInfoAccessor.SetAsync(stepContext.Context, userInfo, cancellationToken);
                    break;

```

This is what I want to replicate but using the OnTurnAsync to do it in between steps. I want to avoid having to go back to the Main Dialog (and ending nested Dialogs to save State Accessor properties. 
By logic, if the OnTurnAsync goes through every turn, this should be possible. But I'm not sure how.
I'm assuming I should use the turnContext, but is there a way to send this info:

```
//TicketNumber is a Class I created as an accessor, similar to Guest Number in the Reuse Dialogs example
((TicketNumber)step.Values[ticketKey]).ticketLetter = ticketLetters;
```
to the OnTurnAsync without ending a dialog and between steps? maybe using:
`return await step.NextAsync()` ?

And i'm also not sure what condition is met every time a turn ends.

However maybe this is not possible?



And, lastly,
So far so long this condition over the OnTurnAsync has never been met:
`if (dialogTurnResult.Status is DialogTurnStatus.Complete)`


Thank you so much for the help.

**Comment by DanteNahuel** [comment]
Another approach I'm trying but not being able to is to Get and Set accessors properties in a Child/Nested Dialog, instead of just over the Main Dialog, using the ReuseDialog documentation examples.
If I could do this, I wouldn't need to use the OnTurnAsync, and instead just save the accessors properties on the go.
However I can't find a way to access main dialogs accessors from the child/nested dialogs, even if I try to send the _accessor object when I add the child dialog to the Main Dialog set.

**Comment by JonathanFingold** [comment]
This is relatively uncharted territory. I'll try two approaches and let you know what my results are:
1. Pass in the ticket number accessor when creating the dialog instances, which will let the dialog update the ticket number in the bot state.
1. Pass in a callback to the bot code when creating the dialog instances, which allows the dialog to pass the ticket number back to the bot code, so the bot can manage state itself.

You could try to examine the DialogState object from the turn handler, but this would probably be unmaintainable in the long run.

**Comment by JonathanFingold** [comment]
Here's a working example with UserState for the ticket property and passing in a callback when creating the dialogs. You could pass in the property accessor directly, but using callbacks allows the bot code to do additional processing. If the conversation ends, and a new conversation starts, the ticket propert will carry forward with the user.

Note that If you don't have to worry about the conversation ending, then you can use the simpler approach, since the dialog "goes away" unexpectedly (Dialog state is kept [generally] in conversation state). One other option to note: if your bot is acting as a front end to another system, your bot is likely making calls to that system to create the ticket and retrieve information about any open tickets, at which point it is not really a bot state issue.

With all those caveats, this should get you going.

Set up some of the classes we'll need elsewhere:
```csharp
public class TicketState
{
    public string Number { get; set; }
}

public class PropertyCallback<T>
{
    public PropertyCallback(Func<ITurnContext, Task<T>> getAsync, Func<ITurnContext, T, Task> setAsync)
    {
        GetPropertyAsync = getAsync;
        SetPropertyAsync = setAsync;
    }

    public Func<ITurnContext, Task<T>> GetPropertyAsync { get; }
    public Func<ITurnContext, T, Task> SetPropertyAsync { get; }
}

public class BotAccessors
{
    public const string DialogStateAccessorName = "DialogState";
    public const string TicketStateAccessorName = "TicketState";

    public BotAccessors(ConversationState conversationState, UserState userState)
    {
        ConversationState = conversationState;
        UserState = userState;
    }

    public ConversationState ConversationState { get; }
    public UserState UserState { get; }

    public IStatePropertyAccessor<DialogState> DialogStateAccessor { get; set; }
    public IStatePropertyAccessor<TicketState> TicketStateAccessor { get; set; }
}
```
Update **Startup.ConfigureServices**. This is how mine looks currently:
```csharp
public void ConfigureServices(IServiceCollection services)
{
    string secretKey = Configuration.GetSection("botFileSecret")?.Value;

    // Loads .bot configuration file and adds a singleton that your Bot can access through dependency injection.
    BotConfiguration botConfig = BotConfiguration.Load(@".\SubComponentDialogs.bot", secretKey);
    services.AddSingleton(sp => botConfig);

    services.AddSingleton(sp =>
    {
        var storage = new MemoryStorage();
        var conversationState = new ConversationState(storage);
        var userState = new UserState(storage);
        return new BotAccessors(conversationState, userState)
        {
            DialogStateAccessor = conversationState.CreateProperty<DialogState>(BotAccessors.DialogStateAccessorName),
            TicketStateAccessor = userState.CreateProperty<TicketState>(BotAccessors.TicketStateAccessorName),
        };
    });

    services.AddBot<SubComponentDialogsBot>(options =>
    {
        // Retrieve current endpoint.
        ConnectedService service = botConfig.Services.Where(s => s.Type == "endpoint" && s.Name == "development").FirstOrDefault();
        if (!(service is EndpointService endpointService))
        {
            throw new InvalidOperationException($"The .bot file does not contain a development endpoint.");
        }

        options.CredentialProvider = new SimpleCredentialProvider(endpointService.AppId, endpointService.AppPassword);

        // Catches any errors that occur during a conversation turn and logs them.
        options.OnTurnError = async (context, exception) =>
        {
            await context.SendActivityAsync("Sorry, it looks like something went wrong:" +
                $"{exception.GetType().Name}: {exception.Message}");
        };
    });
}
```
Setup your dialogs to use the callback object. I've introduced some constants here to help manage the various IDs and strings, to avoid typos. I happen to define steps as lambdas in the first dialog and as methods in the second. In practice, you'd do one or the other consistently.
```csharp
public class ParentDialog : ComponentDialog
{
    private struct DialogIds
    {
        public const string Choice = "choicePrompt";
        public const string MainMenu = "mainDialog";
        public const string OpenATicket = "openDialog";
    }

    private struct Options
    {
        public static readonly Choice OpenATicket = new Choice { Value = "Open a ticket", Synonyms = new List<string> { "open" } };
        public static readonly Choice Exit = new Choice { Value = "Exit ticketing system", Synonyms = new List<string> { "exit" } };
    }

    private static readonly List<Choice> TheChoices = new List<Choice> {
        Options.OpenATicket,
        Options.Exit
    };

    public ParentDialog(string id, PropertyCallback<TicketState> ticketProperty) : base(id)
    {
        InitialDialogId = DialogIds.MainMenu;

        AddDialog(new ChoicePrompt(DialogIds.Choice));
        AddDialog(new ChildDialog(DialogIds.OpenATicket, ticketProperty));

        AddDialog(new WaterfallDialog(DialogIds.MainMenu, new WaterfallStep[] {
            async (stepContext, cancellationToken) =>
            {
                var ticket = await ticketProperty.GetPropertyAsync(stepContext.Context);
                if (ticket?.Number is null)
                {
                    await stepContext.Context.SendActivityAsync("No active ticket...");
                }
                else
                {
                    await stepContext.Context.SendActivityAsync($"Current ticket is `{ticket.Number}`.");
                }

                return await stepContext.PromptAsync(DialogIds.Choice, new PromptOptions
                {
                    Prompt = MessageFactory.Text("Please choose a task."),
                    RetryPrompt = MessageFactory.Text("Please choose one of the options."),
                    Choices = TheChoices,
                });
            },
            async (stepContext, cancellationToken) =>
            {
                FoundChoice choice = stepContext.Result as FoundChoice;
                if (choice.Value == Options.OpenATicket.Value)
                {
                    // Launch the open a ticket dialog.
                    return await stepContext.BeginDialogAsync(DialogIds.OpenATicket, stepContext.Result, cancellationToken);
                }
                // ...
                else // (choice.Value == Options.Exit.Value)
                {
                    // Clear the active ticket info and exit the dialog.
                    await ticketProperty.SetPropertyAsync(stepContext.Context, null);
                    return await stepContext.EndDialogAsync(null, cancellationToken);
                }
            },
            async (stepContext, cancellationToken) =>
            {
                // Loop back to the beginning.
                return await stepContext.ReplaceDialogAsync(DialogIds.MainMenu, null, cancellationToken);
            },
        }));
    }
}

public class ChildDialog : ComponentDialog
{
    private struct DialogIds
    {
        public const string Text = "textPrompt";
        public const string Main = "mainDialog";
    }

    private PropertyCallback<TicketState> TicketProperty { get; }

    public ChildDialog(string id, PropertyCallback<TicketState> ticketProperty) : base(id)
    {
        InitialDialogId = DialogIds.Main;
        TicketProperty = ticketProperty;

        AddDialog(new TextPrompt(DialogIds.Text));

        AddDialog(new WaterfallDialog(DialogIds.Main, new WaterfallStep[] {
            GetTicketNameAsync,
            CreateTicketAsync,
        }));
    }

    private static async Task<DialogTurnResult> GetTicketNameAsync(
        WaterfallStepContext stepContext,
        CancellationToken cancellationToken = default(CancellationToken))
    {
        return await stepContext.PromptAsync(DialogIds.Text, new PromptOptions
        {
            Prompt = MessageFactory.Text("Enter a name for the ticket."),
            RetryPrompt = MessageFactory.Text("Please create a ticket name."),
        });
    }

    private async Task<DialogTurnResult> CreateTicketAsync(
        WaterfallStepContext stepContext,
        CancellationToken cancellationToken = default(CancellationToken))
    {
        string text = stepContext.Result as string;
        await stepContext.Context.SendActivityAsync($"Creating ticket `{text}`...");

        await TicketProperty.SetPropertyAsync(stepContext.Context, new TicketState { Number = text });
        await stepContext.Context.SendActivityAsync($"Ticket `{text}` created.");

        return await stepContext.EndDialogAsync(null, cancellationToken);
    }
}
```
Finally, update your bot code to bring this all together:
```csharp
public class SubComponentDialogsBot : IBot
{
    private BotAccessors _botAccessors;

    private DialogSet _dialogs;

    public SubComponentDialogsBot(BotAccessors botAccessors)
    {
        _botAccessors = botAccessors ?? throw new ArgumentNullException(nameof(botAccessors));
        _dialogs = new DialogSet(_botAccessors.DialogStateAccessor)
            .Add(new ParentDialog("main", new PropertyCallback<TicketState>(GetTicketInfo, SetTicketInfo)));
    }

    private async Task SetTicketInfo(ITurnContext turnContext, TicketState ticketInfo)
    {
        // Perform housekeeping, if necessary.
        await _botAccessors.TicketStateAccessor.SetAsync(turnContext, ticketInfo);
    }

    private async Task<TicketState> GetTicketInfo(ITurnContext turnContext)
    {
        TicketState ticket = await _botAccessors.TicketStateAccessor.GetAsync(turnContext, () => new TicketState());

        // Perform housekeeping, if necessary.
        return ticket;
    }

    public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken = default(CancellationToken))
    {
        // Handle Message activity type, which is the main activity type for shown within a conversational interface
        // Message activities may contain text, speech, interactive cards, and binary or unknown attachments.
        // see https://aka.ms/about-bot-activity-message to learn more about the message and other activity types
        if (turnContext.Activity.Type == ActivityTypes.Message)
        {
            TicketState ticket = await _botAccessors.TicketStateAccessor.GetAsync(turnContext, () => new TicketState());

            DialogContext dc = await _dialogs.CreateContextAsync(turnContext, cancellationToken);
            DialogTurnResult turnResult = await dc.ContinueDialogAsync(cancellationToken);
            switch (turnResult.Status)
            {
                case DialogTurnStatus.Complete:
                    await turnContext.SendActivityAsync("All done.");
                    break;
                case DialogTurnStatus.Empty:
                    await dc.BeginDialogAsync("main");
                    break;
            }

            await _botAccessors.ConversationState.SaveChangesAsync(turnContext);
            await _botAccessors.UserState.SaveChangesAsync(turnContext);
        }
    }
}
```

**Comment by DanteNahuel** [comment]
Thank you for your help
I'm currently trying to implement this into my bot.
I found a couple of issues and doubts, maybe you can help me:

Using this (note that UserInfo = ticketProperty in your code):
`var retries = await UserInfo.GetPropertyAsync(stepContext.Context);`

marks UserInfo **"an object reference is required for the non-static field method or property"**
Strangly, the only way this doesn't happen is if the code is run after a previous (can be even several waterfall steps before)
`return Dialog.EndofTurn; `

I'm not sure what's the logic behind this behavior and if anything else can be done to as a workaround.


Another question:
Lets say I'm using a TicketInfo class with not only int / string properties in it like you showed me, but with subclasses that have properties of their own
E.G:
```
    public class TicketProperty
    {
        public TicketNumber TicketNumber { get; set; }
        public int RetryAttempts { get; set; } = 0;

    }
    public class TicketNumber
    {
        public string ticketNumber { get; set; }
        public string ticketLetter { get; set; }
    
    }
```

How can I set them with the code you showed me?:
`await ticketProperty.SetPropertyAsync(stepContext.Context, new TicketProperty { TicketNumber ??});`
Do I need to convert the TicketProperty class into TicketProperty<T> and pass on only certain types?
or is there a better way to do it?



And, lastly
Lets say I want to use one of the properties as a counter, for example RetryAttemps

`await UserInfo.SetPropertyAsync(stepContext.Context, new UserInfo { RetryAttempts++});`

However, this doesn't work, of course, since RetryAttempts doesn't exist in the context i'm calling it. How can I call this from the UserInfo class?

Thanks in advance.

**Comment by JonathanFingold** [comment]
Some background information: the IPropertyAccessor<T>.GetPropertyAsync method returns a pointer to the object in the turn's state cache, and you can use SetPropertyAsync to write a new object to the cache. If T is a value type, you will need to call SetPropertyAsync any time you want to update the value. If T is a reference type, the call to SetPropertyAsync is optional, as long as you're updating values on the original object. In every case, you need to call the state management object's SaveChangesAsync before the turn ends. This writes any cache changes out to the backing storage layer.

GetPropertyAsync's second parameter is a factory method for generating a new object if one doesn't already exist in the cache. If you don't include this method, you will get somewhat mysterious errors at run time.

The other thing to note in my previous code sample, the _botAccessors property is not static, but I've defined one of the waterfall step methods as static (as it wasn't referencing any instance variables). To simplify things, you can make all the waterfall steps instance methods.

So, within any waterfall step where you need to work with the ticket state, you'd get the object from the cache first:
```csharp
var ticketState = await _botAccessors.TicketAccessor.GetPropertyAsync(stepContext.Context, () => new TicketProperty());
```
Then, update the properties of the ticketState object. These changes will be reflected in the turn's state cache.

**Comment by DanteNahuel** [comment]
Thank you for your answer, you've been very helpful.
I'm guessing I can make all waterfall steps instance methods and remove the static property from them. Since ReplaceDialog steps over the current Dialog, why would you need to make steps Static?
or let me rephrase it, what would happen if I make all steps non-static?

**Comment by JonathanFingold** [comment]
No problem, glad I can help.

Whether to make steps static has to do with how ASP.NET, scaling, and the life-time of the bot. I don't fully understand all the implications. If you set breakpoints in your code, you'll notice that a new instance of your bot is created every turn. Making a bot resource static may reduce processing overhead, if the resource is..._big_ or the bot is handling a lot of traffic.

**Comment by DanteNahuel** [comment]
Something is not working as expected and the property is not being saved as wanted. Obviously i've done something wrong.
This component Dialog I made gets the ticket number from the user, calls a new Dialog to process it and then that Dialog returns an error code depending on whether the ticket exists or not. I'm testing the error route so I can add the retry logic.
I wanted the accessors so I could save the number of retries attempts and access them from all dialogs in the bot.
You can see when the error code returns, I try to add 1 to the counter here:

`await UserInfo.SetPropertyAsync(step.Context, new UserInfo { RetryAttempts =+ 1 });`

However, every time the Dialog restarts, RetryAttempts is still = 1. I'm guessing because i'm calling a new instance of UserInfo every time the Dialog loops?
If that's so, how can I solve this?
If every time the loops the counter accessor is reseted, i'm failing at the logic.

```
using System;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using EchoTest.Dialogs;
using EchoTest.State;

using MarvinServices;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;

public class TicketDialog : ComponentDialog
{
    private const string ticketKey = nameof(TicketDialog);
    private const string TextPrompt = "textPrompt";
    private const string failOver = nameof(FailoverTemp);

    // Next constat is the Id of the current dialog
    private const string ticketDialogComponentId = "ticketComponent";

    // Next constant is the Id of the branched Dialog Secondary Waterfall, required to add the Dialog to the flow.
    private const string ticketProcessingId = "ticketProcessing";

    // Next constant is the Id of the nested child Dialog TicketProcessingDialog, required to add it as a Dialog and call it.
    private const string ticketProcessingDialogId = "ticketprocessingDialog";
    private static UserInfo retry;
    private PropertyCallback<UserInfo> UserInfo { get; }

    // You can start this from the parent using the dialog's ID.
    public TicketDialog(string id, PropertyCallback<UserInfo> userInfo)
        : base(id)
    {
        // The InitialDialogId needs to be set to the ID of a dialog in the nested/child dialog, the one to start when the waterfall starts
        InitialDialogId = ticketDialogComponentId;
        UserInfo = userInfo;
                         

        // Define the prompts used in this conversation flow.
        AddDialog(new TextPrompt(TextPrompt));
        
        // Add the nested Subchild Dialog in this conversation Flow
        AddDialog(new TicketProcessingDialog(ticketProcessingDialogId));


        // Define the conversation flow using a waterfall model.
        WaterfallStep[] step = new WaterfallStep[]
        {
            TicketNumberAsync,
            HandleTicketChoiceAsync,
            FailoverCheckAsync,
            FinalStepAsync,
        };
        AddDialog(new WaterfallDialog(ticketDialogComponentId, step));


//        //This secondary waterfall can be used to branch the dialog with diferent steps, within the same Dialog.cs
//        WaterfallStep[] ticketProcessingSteps = new WaterfallStep[]
//{
//            TicketProcessAsync,
//            TicketReturnAsync,
           
//};
//        AddDialog(new WaterfallDialog(ticketProcessingId, ticketProcessingSteps));

    }

    private async Task<DialogTurnResult> TicketNumberAsync(
        WaterfallStepContext step,
        CancellationToken cancellationToken = default(CancellationToken))
    {
        
        var h = 0;
        step.Values[ticketKey] = new TicketNumber();
        step.Values[failOver] = new FailoverTemp();
        //switch (step.Options)
        //{
        //    case FailoverTemp failover:


        //        break;
        //    case UserInfo userinfo:
        //        // Clear the FailOver property class number and prompt for the ticket number.



        //        break;
        //}



        return await step.PromptAsync(
            TextPrompt,
            new PromptOptions
            {
                Prompt = MessageFactory.Text("Type the ticket letters and numbers (e.g: INCxxxxxxx) : "),

            },
            cancellationToken);
    }

    private static async Task<DialogTurnResult> HandleTicketChoiceAsync(
        WaterfallStepContext step,
        CancellationToken cancellationToken = default(CancellationToken))
    {
        // Save the ticket number as a string
        string resultString = step.Result as string;

       

        //Retrieve the resultString letters and numbers
        string ticketLetters = string.Join("", resultString.ToLower().ToLower().ToCharArray().Where(Char.IsLetter));
        string ticketNumbers = string.Join("", resultString.ToLower().ToCharArray().Where(Char.IsDigit));

        

        //Save ticket number and letters as properties of TicketNumber State. We need to check if this can be used for something, 
        //since we are not returning it to the main dialog and thus it is not saved in the accessor
        ((TicketNumber)step.Values[ticketKey]).ticketLetter = ticketLetters;
        ((TicketNumber)step.Values[ticketKey]).ticketNumber = ticketNumbers;


        // This line calls a branched internal Dialog using the secondary waterfall. After the branched waterfall ends, this Dialog flows continues normally
        //return await step.BeginDialogAsync(ticketProcessingId, i, cancellationToken);

        // This line calls a nested subchild external Dialog. After that nested Dialog is done, this Dialog should continue normally
        return await step.BeginDialogAsync(ticketProcessingDialogId, step.Values[ticketKey], cancellationToken);

               


    }




    private async Task<DialogTurnResult> FailoverCheckAsync(
        WaterfallStepContext step,
        CancellationToken cancellationToken = default(CancellationToken))
    {
        //// Save the room number and "sign off".
        //string letters = step.Result as string;
        //((TicketNumber)step.Values[ticketKey]).ticketLetter = letters;
        
        //string numbers = step.Result as string;
        //((TicketNumber)step.Values[ticketKey]).ticketNumber = numbers;
        var h = 0;

        switch (step.Result)
        {
            case TableInfo table:
 
                
                break;
            //The TicketDialog returns a FailoverTemp object and this case is activated
            case FailoverTemp failovercode:
                // If the user failed to enter an invalid ticket number, TicketprocessingDialog will return a FailOverTemp object and TicketDialog will end, returning a failOver value of one to the main dialog.

                if (failovercode.failOverCode == 1)
                {
                    await step.Context.SendActivityAsync("There's no " + ((TicketNumber)step.Values[ticketKey]).ticketLetter + ((TicketNumber)step.Values[ticketKey]).ticketNumber + " recorded within our Database. Please, enter an existing Ticket");
                    await UserInfo.SetPropertyAsync(step.Context, new UserInfo { RetryAttempts = +1 });
                    return await step.NextAsync();

                }
                else if (failovercode.failOverCode == 2)
                {
                    await step.Context.SendActivityAsync(((TicketNumber)step.Values[ticketKey]).ticketNumber + " doesn't have 7 numbers. Please enter a 7 long incident number");
                    await UserInfo.SetPropertyAsync(step.Context, new UserInfo { RetryAttempts = +1 });
                    return await step.NextAsync();

                }
                else if (failovercode.failOverCode == 3)
                {
                    await step.Context.SendActivityAsync(((TicketNumber)step.Values[ticketKey]).ticketLetter + ((TicketNumber)step.Values[ticketKey]).ticketNumber + " is not in the correct format. Please, enter a 3 letters, 7 numbers ticket");
                    await UserInfo.SetPropertyAsync(step.Context, new UserInfo { RetryAttempts = +1 });
                    return await step.NextAsync();
                }

                else
                {
                    //if the user entered a valid ticket number the TicketDialog should return a FailoverTemp value of 0 and no retryattempt should be logged
                    return await step.ContinueDialogAsync();
                }
                break;

            default:
                // We shouldn't get here, since these are no other branches that get this far.
                break;
        }






        // End the dialog, returning the ticketInfo.
        return await step.EndDialogAsync();
     
    }

    private async Task<DialogTurnResult> FinalStepAsync(
       WaterfallStepContext step,
       CancellationToken cancellationToken = default(CancellationToken))
    {
        retry = await UserInfo.GetPropertyAsync(step.Context);
        if (retry.RetryAttempts > 3)
        {
            return await step.EndDialogAsync(
            (FailoverTemp)step.Values[failOver],
            cancellationToken);
        }
        else
        {
            
            return await step.ReplaceDialogAsync(ticketDialogComponentId, step.Values);
        }

    }

```
